A-blast showcase fails with an assertion

```
godot:master-servo jdm$ ./mach run -d --pref dom.mutation_observer.enabled https://aframe.io/examples/showcase/a-blast/
ERROR 2018-05-17T19:57:36Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
ERROR 2018-05-17T19:57:36Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
ERROR 2018-05-17T19:57:36Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
ERROR 2018-05-17T19:57:36Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
ERROR 2018-05-17T19:57:37Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
ERROR 2018-05-17T19:57:37Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 3) }
ERROR 2018-05-17T19:57:38Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 3) }
ERROR 2018-05-17T19:57:38Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 3) }
ERROR 2018-05-17T19:57:38Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 3) }
A-Frame Version: 0.7.0 (Date 22-09-2017, Commit #570c623e)
three Version:
^0.87.0
WebVR Polyfill Version:
^0.9.36
ERROR 2018-05-17T19:57:42Z: script::dom::bindings::error: Error at https://cdn.jsdelivr.net/docsearch.js/2/docsearch.min.js:2:55856 Error: No input element in the page matches .docs-search-sidebar .search-field
ERROR 2018-05-17T19:57:42Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(8, PipelineId(0, 1)), pipeline_id: PipelineId(1, 3) }
Explosion registered
enemy
Explosion registered
enemygun
Explosion registered
bullet
Explosion registered
background
Bullet registered
default
Bullet registered
enemy-slow
Bullet registered
enemy-medium
Bullet registered
enemy-fast
Bullet registered
enemy-fat
Enemy registered
enemy_start
Enemy registered
enemy0
Enemy registered
enemy1
Enemy registered
enemy2
Enemy registered
enemy3
%ccore:schema:warn %cDefault value `8947848` does not match type `color` in component `collision-helper`%c
color: orange
color: inherit
color: orange
%ccore:schema:warn %cDefault value `1.0` does not match type `number` in component `bmfont-text`%c
color: orange
color: inherit
color: orange
THREE.WebGLRenderer
87
ERROR 2018-05-17T19:57:43Z: offscreen_gl_context::draw_buffer: The given GLContext doesn't support requested antialising
THREE.WebGLRenderer: WEBGL_depth_texture extension not supported.
THREE.WebGLRenderer: ANGLE_instanced_arrays extension not supported.
THREE.WebGLRenderer: OES_element_index_uint extension not supported.
ERROR 2018-05-17T19:57:43Z: offscreen_gl_context::draw_buffer: The given GLContext doesn't support requested antialising
ERROR 2018-05-17T19:57:43Z: offscreen_gl_context::draw_buffer: The given GLContext doesn't support requested antialising
%ccore:a-assets:warn %cAsset loading timed out in %c
color: orange
color: inherit
color: orange
3000
ms
%ccomponents:sound:warn %cAudio source was not specified with `src`%c
color: orange
color: inherit
color: orange
assertion failed: `(left == right)`
  left: `Receiving([...])`,
 right: `Empty` (thread fetch thread for https://aframe.io/a-blast/assets/models/gun.json, at components/net/fetch/methods.rs:445)
stack backtrace:
   0:        0x114dace34 - backtrace::backtrace::trace::he39332bd6b104fb9
                        at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.2/src/backtrace/mod.rs:42
   1:        0x114da7f5c - backtrace::capture::Backtrace::new::h56fe28cf6edb3743
                        at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.2/src/capture.rs:64
   2:        0x10df98f7d - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h07036b9f5999b66e
                        at ports/servo/main.rs:145
   3:        0x114de3934 - std::panicking::rust_panic_with_hook::hcebe05e8466ff134
                        at libstd/panicking.rs:467
   4:        0x114de3420 - std::panicking::begin_panic_fmt::hf6867449b867ca07
                        at libstd/panicking.rs:350
   5:        0x11170ebf4 - net::fetch::methods::wait_for_response::h269c2ffba5b52992
                        at components/net/fetch/methods.rs:445
   6:        0x11170d382 - net::fetch::methods::main_fetch::h7c217cd5f43be2dd
                        at components/net/fetch/methods.rs:405
   7:        0x11170ba4c - net::fetch::methods::fetch_with_cors_cache::h9102f7e05e1bdf72
                        at components/net/fetch/methods.rs:124
   8:        0x11170b8c1 - net::fetch::methods::fetch::h567a6e669ce2fcc3
                        at components/net/fetch/methods.rs:86
   9:        0x111535983 - net::resource_thread::CoreResourceManager::fetch::_$u7b$$u7b$closure$u7d$$u7d$::hcf7f1ab5ad3295b3
                        at components/net/resource_thread.rs:435
  10:        0x1116d4054 - std::sys_common::backtrace::__rust_begin_short_backtrace::h2a52bc61f575bfd7
                        at /Users/travis/build/rust-lang/rust/src/libstd/sys_common/backtrace.rs:136
  11:        0x1115e5c0b - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h3f5a16d6e8eab95f
                        at /Users/travis/build/rust-lang/rust/src/libstd/thread/mod.rs:409
  12:        0x1116d3364 - _$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h3dbd55ab1a787b6f
                        at /Users/travis/build/rust-lang/rust/src/libstd/panic.rs:305
  13:        0x1116d73f4 - std::panicking::try::do_call::h1d00b7c2249a9d63
                        at /Users/travis/build/rust-lang/rust/src/libstd/panicking.rs:310
  14:        0x114def11e - __rust_maybe_catch_panic
                        at libpanic_unwind/lib.rs:105
  15:        0x1116d6f49 - std::panicking::try::he1cbcc1b3a140e7e
                        at /Users/travis/build/rust-lang/rust/src/libstd/panicking.rs:289
  16:        0x1116d5b04 - std::panic::catch_unwind::h6ff731804fddebcf
                        at /Users/travis/build/rust-lang/rust/src/libstd/panic.rs:374
  17:        0x1115e4d08 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::h82d37769556dc11a
                        at /Users/travis/build/rust-lang/rust/src/libstd/thread/mod.rs:408
  18:        0x1115e70e5 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::ha3a3e30e457aa6e4
                        at /Users/travis/build/rust-lang/rust/src/liballoc/boxed.rs:638
  19:        0x114ddee97 - std::sys_common::thread::start_thread::h3a2a1586806f3926
                        at libstd/sys_common/thread.rs:24
  20:        0x114dc7d18 - std::sys::unix::thread::Thread::new::thread_start::h657bd8405a5d884c
                        at libstd/sys/unix/thread.rs:90
  21:     0x7fffbfb9093a - _pthread_body
  22:     0x7fffbfb90886 - _pthread_start
```

--- a/components/net/http_cache.rs
+++ b/components/net/http_cache.rs
@@ -624,6 +624,23 @@ impl HttpCache {
         let entry_key = CacheKey::new(request.clone());
         if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
             for cached_resource in cached_resources.iter_mut() {
+                // done_chan will have been set to Some(..) by http_network_fetch.
+                // If the body is not receiving data, set the done_chan back to None.
+                // Otherwise, create a new dedicated channel to update the consumer.
+                // The response constructed here will replace the 304 one from the network.
+                let in_progress_channel = match *cached_resource.body.lock().unwrap() {
+                    ResponseBody::Receiving(..) => {
+                        Some(channel())
+                    },
+                    ResponseBody::Empty | ResponseBody::Done(..) => None
+                };
+                match in_progress_channel {
+                    Some((done_sender, done_receiver)) => {
+                        *done_chan = Some((done_sender.clone(), done_receiver));
+                        cached_resource.awaiting_body.lock().unwrap().push(done_sender);
+                    },
+                    None => *done_chan = None
+                }
                 // Received a response with 304 status code, in response to a request that matches a cached resource.
                 // 1. update the headers of the cached resource.
                 // 2. return a response, constructed from the cached resource.
@@ -635,9 +652,6 @@ impl HttpCache {
                 constructed_response.referrer_policy = request.referrer_policy.clone();
                 constructed_response.raw_status = cached_resource.data.raw_status.clone();
                 constructed_response.url_list = cached_resource.data.url_list.clone();
-                // done_chan will have been set to Some by http_network_fetch,
-                // set it back to None since the response returned here replaces the 304 one from the network.
-                *done_chan = None;
                 cached_resource.data.expires = get_response_expiry(&constructed_response);
                 let mut stored_headers = cached_resource.data.metadata.headers.lock().unwrap();
                 stored_headers.extend(response.headers.iter());

--- a/components/net/http_loader.rs
+++ b/components/net/http_loader.rs
@@ -905,26 +905,30 @@ fn http_network_or_cache_fetch(request: &mut Request,
         }
     }
 
-    if let Some(ref ch) = *done_chan {
-        // The cache constructed a response with a body of ResponseBody::Receiving.
-        // We wait for the response in the cache to "finish",
-        // with a body of either Done or Cancelled.
-        loop {
-            match ch.1.recv()
-                    .expect("HTTP cache should always send Done or Cancelled") {
-                Data::Payload(_) => {},
-                Data::Done => break, // Return the full response as if it was initially cached as such.
-                Data::Cancelled => {
-                    // The response was cancelled while the fetch was ongoing.
-                    // Set response to None, which will trigger a network fetch below.
-                    response = None;
-                    break;
+    fn wait_for_cached_response(done_chan: &mut DoneChannel, response: &mut Option<Response>) {
+        if let Some(ref ch) = *done_chan {
+            // The cache constructed a response with a body of ResponseBody::Receiving.
+            // We wait for the response in the cache to "finish",
+            // with a body of either Done or Cancelled.
+            loop {
+                match ch.1.recv()
+                        .expect("HTTP cache should always send Done or Cancelled") {
+                    Data::Payload(_) => {},
+                    Data::Done => break, // Return the full response as if it was initially cached as such.
+                    Data::Cancelled => {
+                        // The response was cancelled while the fetch was ongoing.
+                        // Set response to None, which will trigger a network fetch below.
+                        *response = None;
+                        break;
+                    }
                 }
             }
         }
+        // Set done_chan back to None, it's cache-related usefulness ends here.
+        *done_chan = None;
     }
-    // Set done_chan back to None, it's cache-related usefulness ends here.
-    *done_chan = None;
+
+    wait_for_cached_response(done_chan, &mut response);
 
     // Step 22
     if response.is_none() {
@@ -951,6 +955,7 @@ fn http_network_or_cache_fetch(request: &mut Request,
         if revalidating_flag && forward_response.status.map_or(false, |s| s == StatusCode::NotModified) {
             if let Ok(mut http_cache) = context.state.http_cache.write() {
                 response = http_cache.refresh(&http_request, forward_response.clone(), done_chan);
+                wait_for_cached_response(done_chan, &mut response);
             }
         }
 

--- a/components/net/tests/http_cache.rs
+++ b/components/net/tests/http_cache.rs
@@ -1,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use hyper::header::{Expires, HttpDate};
+use hyper::method::Method;
+use hyper::status::StatusCode;
+use msg::constellation_msg::TEST_PIPELINE_ID;
+use net::http_cache::HttpCache;
+use net_traits::request::{Destination, Request, RequestInit};
+use net_traits::response::{Response, ResponseBody};
+use servo_url::ServoUrl;
+use std::sync::mpsc::channel;
+use time;
+
+
+#[test]
+fn test_refreshing_resource_sets_done_chan_the_appropriate_value() {
+    let response_bodies = vec![ResponseBody::Receiving(vec![]),
+                               ResponseBody::Empty,
+                               ResponseBody::Done(vec![])];
+    let url = ServoUrl::parse("https://servo.org").unwrap();
+    let request = Request::from_init(RequestInit {
+        url: url.clone(),
+        method: Method::Get,
+        destination: Destination::Document,
+        origin: url.clone().origin(),
+        pipeline_id: Some(TEST_PIPELINE_ID),
+        .. RequestInit::default()
+    });
+    let mut response = Response::new(url.clone());
+    // Expires header makes the response cacheable.
+    response.headers.set(Expires(HttpDate(time::now())));
+    response_bodies.iter().for_each(|body| {
+        let mut cache = HttpCache::new();
+        *response.body.lock().unwrap() = body;
+        // First, store the 'normal' response.
+        cache.store(&request, &response);
+        // Second, mutate the response into a 304 response, and refresh the stored one.
+        response.status = Some(StatusCode::NotModified);
+        let mut done_chan = Some(channel());
+        let refreshed_response = cache.refresh(&request, response, &mut done_chan);
+        // Ensure a resource was found, and refreshed.
+        assert!(refreshed_response.is_some());
+        match body {
+            ResponseBody::Receiving(_) => assert!(done_chan.is_some()),
+            ResponseBody::Empty | ResponseBody::Done(_) => assert!(done_chan.is_none())
+        }
+    })
+}

